### PR TITLE
Fix for CosmosDB race conditions that resulted in intermittent 500s

### PR DESCRIPTION
**Pull request by moritzraho**
## Description
Tested manually via:
- Requires to set .env credentials first
- Requires to delete the permission user attached to the namespace first: `await database.user(user-${Buffer.from(owNamespace).toString('hex')).delete()}`

- run 10 concurrent requests to trigger race conditions
```
for ((i=1;i<10;i++)) ;                                    
do nohup node scripts/run azure-cosmos owNamespace=<ns> __ow_headers='{ "authorization": "Basic <Buffer.from(auth).toString('base64')>" }' &
done
```

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
- [x] My code follows the code style of this project.
- [] My change requires a change to the documentation.
- [] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: lib/impl/AzureCosmosTvm.js [hunk 1]
```diff
@@ -77,6 +96,46 @@ class AzureCosmosTvm extends Tvm {
       partitionKey
     }
   }
+
+  async createUserAndPermission (/** @type {cosmos.Database} */ database, userId, permissionDefinition, resourceTokenExpirySeconds) {
```
Comment (purplecabbage) [inline lib/impl/AzureCosmosTvm.js:100]: since this function only needs the `id` why not just pass that in?

File: lib/impl/AzureCosmosTvm.js [hunk 2]
```diff
@@ -77,6 +96,46 @@ class AzureCosmosTvm extends Tvm {
...
+      }
+
+      logger.info('detected conflict, entering race condition resolution')
+
+      // retry 5 times with 50 msec delay
+      for (let retryCount = 1; retryCount <= RACE_RETRIES; ++retryCount) {
+        try {
+          // attempt to retrieve the already created permission
+          logger.info(`get permission '${permissionDefinition.id}'`)
+          const permissionRes = await database.user(userId)
```
Comment (purplecabbage) [inline lib/impl/AzureCosmosTvm.js:121]: How many async calls is this? Are they ALL awaited?

**Review by purplecabbage — APPROVED** [review]
This to me doesn't seem to resolve the race condition, just retry and hope it does not happen next time.
What do we know about the race condition, and the cause?
If this is a network issue, then we should be using exponential fallback, instead of 50ms
50, 100, 200, 400, 800, etc ..

File: lib/impl/AzureCosmosTvm.js (hunk 1 above)
Comment (moritzraho) [reply lib/impl/AzureCosmosTvm.js:100]: it needs the cosmos db client (database) + userId + permission creation payload (which includes the id) + expiry

**Comment by moritzraho** [comment]
> This to me doesn't seem to resolve the race condition, just retry and hope it does not happen next time.
What do we know about the race condition, and the cause?
If this is a network issue, then we should be using exponential fallback, instead of 50ms
50, 100, 200, 400, 800, etc ..

Sorry for missing out on the details, now the race condition can only happen when the user accesses state lib for the first time via multiple concurrent requests (before it could happen on every concurrent request as we where doing a delete + create):
- request a and request b both try to read the permission
- the permission does not exist, so request a and b need to create the user + permission
- let's say request a creates the resources first and returns with the credentials
- request b will attempt to create the user and get 409
- so it to resolve the race condition it then assumes that another request (a.k.a request a) has created the resource already, so it simply attempt to get the resource
- retries are just here to wait for the resource to be available in cosmosdb

**Comment by purplecabbage** [comment]
or 1.61803398875 golden ratio factor

File: lib/impl/AzureCosmosTvm.js (hunk 1 above)
Comment (purplecabbage) [reply lib/impl/AzureCosmosTvm.js:100]: permissionDefinition.id that is ...

**Pull request merged**